splitgpg2/__init__.py

```diff
@@ -819,44 +838,51 @@ def escape_D(data):
     # we send the reserialized form this should be safe.
 
     @classmethod
-    def parse_sexpr(klass, untrusted_arg):
+    def parse_sexpr(klass, untrusted_arg: bytes):
+        if type(untrusted_arg) is not bytes:
+            raise TypeError("invalid type in parse_sexpr")
         if len(untrusted_arg) == 0:
             raise ValueError("no sexpr")
+        if untrusted_arg[0] != ord('('):
+            raise ValueError("sexpr must start with (")
         sexpr, rest = klass._parse_sexpr(untrusted_arg, 0)
         if len(rest) != 0:
             raise ValueError("garbage at end of sexpr")
         if len(sexpr) != 1:
-            raise ValueError("sexpr top level shold have exactly one element")
+            raise ValueError("sexpr top level shold have exactly one element, not %d" % len(sexpr))
         return sexpr[0]
 
     @classmethod
-    def _parse_sexpr(klass, untrusted_arg, nesting):
-        if len(untrusted_arg) == 0:
-            if nesting > 0:
-                raise ValueError("missing closing parenthesis")
+    def _parse_sexpr(klass, untrusted_arg, depth, parens=0):
+        if depth > 10:
+            raise ValueError("nesting depth too deep")
+        if not untrusted_arg:
+            if parens != 0:
+                raise ValueError("Missing close parentheses")
             return ([], b'')
         elif untrusted_arg[0] == ord(')'):
-            if nesting == 0:
-                return ([], untrusted_arg)
+            if parens < 1:
+                raise ValueError("Unexpected close paremtheses")
             return ([], untrusted_arg[1:].lstrip(b' '))
 
-        if untrusted_arg[0] in range(0x30, 0x40):
+        if 0x30 <= untrusted_arg[0] < 0x40:
```

Comment:
I like the `range` better (not a fan of "chained comparisons"). Does this have an real advantage? Please keep separate changes in separate commits.

Reply:
It does not.